### PR TITLE
Various cleanups around the const eval query providers

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung 

after this, working on running validation before interning starts with swapping the order of two lines of code

### CODE REVIEW

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 1]
```diff
@@ -282,18 +285,23 @@ pub fn eval_static_initializer_provider<'tcx>(
...
-    );
-    let alloc_id = eval_in_interpreter(&mut ecx, cid, true)?.alloc_id;
-    let alloc = take_static_root_alloc(&mut ecx, alloc_id);
-    let alloc = tcx.mk_const_alloc(alloc);
-    Ok(alloc)
+    eval_in_interpreter(tcx, cid, ty::ParamEnv::reveal_all())
+}
+
+pub trait InterpretationResult<'tcx> {
+    fn make_result<'mir>(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:292]: This should come with a doc comment. In particular, this shouldn't be called during evaluation, only after the end.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 2]
```diff
@@ -319,92 +327,64 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
...
-                },
-            ))
-        }
-        Ok(mplace) => {
-            // Since evaluation had no errors, validate the resulting constant.
-
-            // Temporarily allow access to the static_root_ids for the purpose of validation.
-            let static_root_ids = ecx.machine.static_root_ids.take();
-            let res = const_validate_mplace(&ecx, &mplace, cid);
-            ecx.machine.static_root_ids = static_root_ids;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:387]: What is going on with the diff here -- I changed this code in https://github.com/rust-lang/rust/pull/122249, didn't I?

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 3]
```diff
@@ -418,7 +398,8 @@ pub fn const_validate_mplace<'mir, 'tcx>(
                 CtfeValidationMode::Const { allow_immutable_unsafe_cell: !inner }
             }
         };
-        ecx.const_validate_operand(&mplace.into(), path, &mut ref_tracking, mode)?;
+        ecx.const_validate_operand(&mplace.into(), path, &mut ref_tracking, mode)
+            .map_err(|error| const_report_error(&ecx, error, alloc_id))?;
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:402]: `const_report_error` seems to be specifically for reporting *validation* failures -- might be better to rename it to `report_validation_error` and make it private?

Also would be good to explain here why we don't just forward the `InterpError`: to give a more targeted error when final validation fails.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 4]
```diff
@@ -444,7 +425,7 @@ pub fn const_report_error<'mir, 'tcx>(
         *ecx.tcx,
         error,
         None,
-        || crate::const_eval::get_span_and_frames(ecx.tcx, &ecx.machine),
+        || crate::const_eval::get_span_and_frames(ecx.tcx, ecx.stack()),
         move |span, frames| errors::UndefinedBehavior { span, ub_note, frames, raw_bytes },
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:429]: `errors::UndefinedBehavior` is poorly named, this is specifically for validation failure, not all const-eval UB.

**Review by RalfJung — COMMENTED** [review]
Thanks, looks good!

Some nits and then I saw some chances for further cleanup. Feel free to disregard the latter if you don't want to expand the scope of this PR.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:387]: Oh, you brought this back in https://github.com/rust-lang/rust/pull/121644. Was that deliberate?

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:387]: Probably a rebase screwup. I thought I rebased correctly, apparently not

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:387]: I made it an explicit commit at the start of the PR

File: compiler/rustc_const_eval/src/errors.rs [hunk 5]
```diff
@@ -410,7 +410,7 @@ pub struct NullaryIntrinsicError {
 
 #[derive(Diagnostic)]
 #[diag(const_eval_undefined_behavior, code = E0080)]
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/errors.rs:412]: The fluent constant still has the confusing old name.

**Pull request merged**